File: clojure/tp/recursion.clj
(ns clojure.examples.hello
  (:gen-class))

; counting with recursion
(defn count [to]
  (loop [i 0]
    (when (< i to)
      (println i)
      (recur (inc i)))))


; mandatory fib recursion example

; slow version (from rosetta code)
(defn slow-fib [n]
  (case n
    0 0
    1 1
    (+ (slow-fib (- n 1))
       (slow-fib (- n 2)))))

; memoized (from rosetta code)
(def mem-fib
  (memoize
    (fn [n]
      (case n
        0 0
        1 1 
        (+ (mem-fib (- n 1))
           (mem-fib (- n 2)))))))

; fast with the "doubling algorithm" (rosetta code)
(defn fast-fib [n]
  (letfn [(fast-fib* [n]
            (if (zero? n)
              [0 1]
              (let [[a b] (fast-fib* (quot n 2))
                    c (*' a (-' (*' 2 b) a))
                    d (+' (*' b b) (*' a a))]
                (if (even? n)
                  [c d]
                  [d (+' c d)]))))]
    (first (fast-fib* n))))

(count 10)

(println "naive fib vs memoized naive fib vs fast fibonacci")

(def x 30)

(println "naive:    " (time (slow-fib x)))
(println "memoized: " (time (mem-fib  x)))
(println "fast:     " (time (fast-fib x)))

File: erlang/tp/recursion.erl
-module(recursion).
-export([while/1, while/2, start/0, fact/1, list_length/1]).
-on_load(start/0).

while(L) -> while(L, 0).
while([], Acc) -> Acc;

while([_|T], Acc) ->
  io:fwrite("~w~n", [Acc]),
  while(T, Acc+1).

for(0, _) ->
  [];

  for(N, Term) when N > 0 ->
    io:fwrite("Hello~n"),
    [Term|for(N-1, Term)].

% mandatory factorial example
fact(N) when N == 0 -> 1;
fact(N) when N > 0 -> N * fact(N - 1).

% list length
list_length([]) -> 0;
list_length([_|T]) -> 1 + list_length(T).


start() ->
  X = [1, 2, 3, 4, 5],
  while(X),
  for(5,1),

  io:fwrite("Mandatory factorial example~n"),
  io:fwrite("factorial of 4 is ~w~n", [fact(4)]),

  io:fwrite("the length of list X is ~w~n", [list_length(X)]).

